Git Token Based Authentication Documentation (#221)
nephio-project/docs: content/en/docs/porch/user-guides/git-authentication-config.md

---
title: "Git Authentication Configuration"
type: docs
weight: 4
description: ""
---

## 1. Porch Server to Git Interaction

The porch server handles interaction with associated git repositories through the use of porch repository CR (Custom Resource) which act as a link between the porch server and the git repositories the server is meant to interact with and store packages on.

> More information on porch repositories can be found [here](https://docs.nephio.org/docs/porch/package-orchestration/#repositories).

Reviewer: ```suggestion
More information on porch repositories can be found [here](../package-orchestration.md#repositories).
```

Author: ah right i knew i could use the local file structure instead of URL's but i could not remember how. Thank you.

### 1.1 Porch Repository Basic Authentication Configuration

> An porch repository object can be created through the use of the `porchctl repo reg porch-test-repository -n porch-test http://example-ip:example-port/repo.git --repo-basic-password=password --repo-basic-username=username` command which creates a secret and repository object.

Reviewer 2: ```suggestion
A porch repository object can be created through the use of the `porchctl repo reg porch-test-repository -n porch-test http://example-ip:example-port/repo.git --repo-basic-password=password --repo-basic-username=username` command which creates a secret and repository object.
```

Author: amended

type: Opaque
```

{{% alert title="Note" color="primary" %}}
The Porch server caches the authentication credentials from the secret, so if the secret's contents are updated they may in fact not be the credentials which are used in the authentication until the old secret credentials are no longer valid which triggers the porch server to query the secret again use the new credentials which if valid become the new cached authentication values.
{{% /alert %}}

### 1.3 Porch Repository HTTPS Configuration

To enable the porch server to communicate with a custom git deployment over HTTPS, we must:

1. Provide an additional arguments flag `use-git-cabundle=true` to the porch-server deployment.
2. Provide an additional Kubernetes secret containing the relevant certificate chain in the form of a cabundle.

The secret itself must meet the following criteria:

- Exist in the same namespace as the Repository CR that requires it.
- Be named specifically \<namespace\>-ca-bundle.
- Have a Data key named *ca.crt* containing the relevant ca certificate (chain).

For example, a Git Repository is hosted over HTTPS at the *https://my-gitlab.com/joe.bloggs/blueprints.git* URL:

> `https://my-gitlab.com/joe.bloggs/blueprints.git`

Author: my apologies. this section is from the previous documentation which was on this page before my changes, i presumed that since it was inside the documentation already it was fine as is and left it mostly untouched. ill make amends as suggested.

> `https://my-gitlab.com/joe.bloggs/blueprints.git`

Author: amended

---

> title: "Git Authentication Configuration"

Reviewer 3: Authenticating to Remote Git Repositories

Author: amended

type: docs
weight: 4
description: ""
---

> ## 1. Porch Server to Git Interaction

Reviewer 3: I think we can drop the numbered headings

Author: amended

The porch server handles interaction with associated git repositories through the use of porch repository CR (Custom Resource) which act as a link between the porch server and the git repositories the server is meant to interact with and store packages on.

More information on porch repositories can be found [here](../package-orchestration.md#repositories).

> 

Reviewer 3: Add very short blurb on what follows..ie. ways to authenticate towards git..

Author: amended

> ### 1.1 Porch Repository Basic Authentication Configuration

Reviewer 3: ```suggestion
### Basic Authentication
```

Author: amended

- Exist in the same namespace as the Repository CR (Custom Resource) that requires it.
- Have a Data keys named *username* and *password* containing the relevant information.
- Be of type *basic-auth*.

The value used in the *password* field can be substituted for a base64 encoded Personal Access Token (PAT) from the GIT instance being used. An Example of this can be found [here](./porchctl-cli-guide.md#repository-registration)

Which would be the equivalent of doing a `kubectl apply -f` on a yaml file with the following content (assuming the porch-test namespace exists on the cluster):

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: git-auth-secret
  namespace: porch-test
data:
  username: base-64-encoded-username
  password: base-64-encoded-password  # or base64-encoded-PAT
type: kubernetes.io/basic-auth

---
apiVersion: config.porch.kpt.dev/v1alpha1
kind: Repository

metadata:
  name: porch-test-repository
  namespace: porch-test

spec:
  description: porch test repository
  content: Package
  deployment: false
  type: git
  git:
    repo: http://example-ip:example-port/repo.git
    directory: /
    branch: main
    secretRef:
      name: git-auth-secret
```

> When The Porch Server is interacting with a Git instance through this http-basic-auth configuration it does so through HTTP. An example HTTP Request using this configuration can be seen below.

Author: amended

```logs

> PUT

Reviewer 3: You sure this is a PUT?

Author: i believe so. i exported this example directly from a log in the porch-server but i could be wrong.

https://example-ip/apis/config.porch.kpt.dev/v1alpha1/namespaces/porch-test/repositories/porch-test-repo/status
Request Headers:
     User-Agent: __debug_bin1520795790/v0.0.0 (linux/amd64) kubernetes/$Format
     Authorization: Basic bmVwaGlvOnNlY3JldA== 
     Accept: application/json, */*
     Content-Type: application/json
```

> where *bmVwaGlvOnNlY3JldA==* is base64 encoded in the format *username:password* and after base64 decoding becomes *nephio:secret* and for simple personal access token login the password section can be substituted for the PAT token.

Reviewer 3: ```suggestion
where *bmVwaGlvOnNlY3JldA==* is base64 encoded in the format *username:password* and after base64 decoding becomes *nephio:secret*. For simple personal access token login, the password section can be substituted with the PAT token.
```

Author: amended

> ### 1.2 Porch Repository Bearer Token Authentication Configuration

Reviewer 3: ```suggestion
### Bearer Token Authentication
```

Author: amended

The authentication to the git repository can be configured to be in bearer token format by altering the secret used in the porch repository object.

The bearer token authentication secret must meet the following criteria:

- Exist in the same namespace as the Repository CR (Custom Resource) that requires it
- Have a Data key named *bearerToken* containing the relevant git token information.
- Be of type *Opaque*.

For example:

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: git-auth-secret
  namespace: porch-test
data:
  bearerToken: base-64-encoded-bearer-token
type: Opaque
```

> When The Porch Server is interacting with a Git instance through this http-token-auth configuration it does so through HTTP. An example HTTP Request using this configuration can be seen below.

Author: amended

```logs
PUT https://example-ip/apis/config.porch.kpt.dev/v1alpha1/namespaces/porch-test/repositories/porch-test-repo/status
Request Headers:
     User-Agent: __debug_bin1520795790/v0.0.0 (linux/amd64) kubernetes/$Format
     Authorization: Bearer 4764aacf8cc6d72cab58e96ad6fd3e3746648655
     Accept: application/json, */*
     Content-Type: application/json
```

> where *4764aacf8cc6d72cab58e96ad6fd3e3746648655* in the Authorization header is a PAT token but can be whichever type of bearer token is accepted by the user's git instance.

Reviewer 3: ```suggestion
where *4764aacf8cc6d72cab58e96ad6fd3e3746648655* in the Authorization header is a PAT token, but can be whichever type of bearer token is accepted by the user's git instance.
```

Author: amended

{{% alert title="Note" color="primary" %}}

> The Porch server caches the authentication credentials from the secret, so if the secret's contents are updated they may in fact not be the credentials which are used in the authentication until the old secret credentials are no longer valid which triggers the porch server to query the secret again use the new credentials which if valid become the new cached authentication values.

Reviewer 3: Break this up into a few sentences

Author: done

{{% /alert %}}

> ### 1.3 Porch Repository HTTPS Configuration

Reviewer 3: ```suggestion
### HTTPS/TLS Configuration
```

Author: amended
